### ISSUE TITLE
Failed to mock entity with composite key in Grails 2.5.4

### ISSUE THREAD

**Issue by badbob** [grails/grails-core#10488]
```
org.grails.datastore.mapping.model.IllegalMappingException: Mapped identifier [id] for class [foo.bar.model.security.UserRole] is not a valid property
	at org.grails.datastore.mapping.model.config.GormMappingConfigurationStrategy.getIdentity(GormMappingConfigurationStrategy.java:807)
	at org.grails.datastore.mapping.model.AbstractPersistentEntity.resolveIdentifier(AbstractPersistentEntity.java:125)
	at org.grails.datastore.mapping.model.AbstractPersistentEntity.initialize(AbstractPersistentEntity.java:88)
	at org.grails.datastore.mapping.model.AbstractMappingContext.initializePersistentEntity(AbstractMappingContext.java:250)
	at org.grails.datastore.mapping.model.AbstractMappingContext.addPersistentEntities(AbstractMappingContext.java:183)
	at grails.test.mixin.domain.DomainClassUnitTestMixin.mockDomains(DomainClassUnitTestMixin.groovy:86)
	at org.spockframework.util.ReflectionUtil.invokeMethod(ReflectionUtil.java:153)
	at org.spockframework.runtime.model.MethodInfo.invoke(MethodInfo.java:84)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:88)
	at org.spockframework.runtime.extension.builtin.AbstractRuleInterceptor$1.evaluate(AbstractRuleInterceptor.java:37)
	at grails.test.runtime.TestRuntimeJunitAdapter$1$2.evaluate(TestRuntimeJunitAdapter.groovy:49)
	at org.spockframework.runtime.extension.builtin.TestRuleInterceptor.intercept(TestRuleInterceptor.java:38)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:87)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:88)
	at org.spockframework.runtime.extension.builtin.AbstractRuleInterceptor$1.evaluate(AbstractRuleInterceptor.java:37)
	at grails.test.runtime.TestRuntimeJunitAdapter$3$4.evaluate(TestRuntimeJunitAdapter.groovy:76)
	at org.spockframework.runtime.extension.builtin.ClassRuleInterceptor.intercept(ClassRuleInterceptor.java:38)
	at org.spockframework.runtime.extension.MethodInvocation.proceed(MethodInvocation.java:87)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.codehaus.groovy.grails.test.spock.GrailsSpecTestType.doRun(GrailsSpecTestType.groovy:147)
	at org.codehaus.groovy.grails.test.support.GrailsTestTypeSupport.run(GrailsTestTypeSupport.groovy:121)
	at org.codehaus.groovy.grails.test.runner.GrailsProjectTestRunner.runTests(GrailsProjectTestRunner.groovy:415)
	at org.codehaus.groovy.grails.test.runner.GrailsProjectTestRunner.processTests(GrailsProjectTestRunner.groovy:395)
	at org.codehaus.groovy.grails.test.runner.GrailsProjectTestRunner.runAllTests_closure7(GrailsProjectTestRunner.groovy:314)
	at groovy.lang.Closure.call(Closure.java:426)
	at org.codehaus.groovy.grails.test.runner.GrailsProjectTestRunner.runAllTests(GrailsProjectTestRunner.groovy:300)
	at org.codehaus.groovy.grails.test.runner.GrailsProjectTestRunner.runAllTests(GrailsProjectTestRunner.groovy:215)
	at _GrailsTest.run_closure1(_GrailsTest.groovy:102)
	at org.codehaus.gant.GantMetaClass.invokeMethod(GantMetaClass.java:133)
	at org.codehaus.gant.GantBinding.initializeGantBinding_closure5_closure16_closure18(GantBinding.groovy:185)
	at org.codehaus.gant.GantBinding.initializeGantBinding_closure5_closure16_closure18(GantBinding.groovy)
	at org.codehaus.gant.GantBinding.withTargetEvent(GantBinding.groovy:90)
	at groovy.lang.GroovyObjectSupport.invokeMethod(GroovyObjectSupport.java:46)
	at org.codehaus.gant.GantBinding.initializeGantBinding_closure5_closure16(GantBinding.groovy:185)
	at org.codehaus.gant.GantBinding.initializeGantBinding_closure5_closure16(GantBinding.groovy)
	at groovy.lang.GroovyObjectSupport.invokeMethod(GroovyObjectSupport.java:46)
	at groovy.lang.Script.invokeMethod(Script.java:80)
	at org.codehaus.gant.GantMetaClass.invokeMethod(GantMetaClass.java:133)
	at TestApp.run_closure1(TestApp.groovy:32)
	at org.codehaus.gant.GantMetaClass.invokeMethod(GantMetaClass.java:133)
	at org.codehaus.gant.GantBinding.initializeGantBinding_closure5_closure16_closure18(GantBinding.groovy:185)
	at org.codehaus.gant.GantBinding.initializeGantBinding_closure5_closure16_closure18(GantBinding.groovy)
	at org.codehaus.gant.GantBinding.withTargetEvent(GantBinding.groovy:90)
	at groovy.lang.GroovyObjectSupport.invokeMethod(GroovyObjectSupport.java:46)
	at org.codehaus.gant.GantBinding.initializeGantBinding_closure5_closure16(GantBinding.groovy:185)
	at org.codehaus.gant.GantBinding.initializeGantBinding_closure5_closure16(GantBinding.groovy)
	at gant.Gant.dispatch_closure5(Gant.groovy:381)
	at gant.Gant.dispatch_closure7(Gant.groovy:415)
	at gant.Gant.dispatch_closure7(Gant.groovy)
	at gant.Gant.withBuildListeners(Gant.groovy:427)
	at gant.Gant.dispatch(Gant.groovy:415)
	at gant.Gant.executeTargets(Gant.groovy:591)
	at gant.Gant.executeTargets(Gant.groovy:590)
	at org.codehaus.groovy.grails.cli.GrailsScriptRunner.executeWithGantInstance(GrailsScriptRunner.java:779)
	at org.codehaus.groovy.grails.cli.GrailsScriptRunner.executeScriptFile(GrailsScriptRunner.java:571)
	at org.codehaus.groovy.grails.cli.GrailsScriptRunner.executeScriptWithCaching(GrailsScriptRunner.java:470)
	at org.codehaus.groovy.grails.cli.GrailsScriptRunner.callPluginOrGrailsScript(GrailsScriptRunner.java:414)
	at org.codehaus.groovy.grails.cli.GrailsScriptRunner.executeCommand(GrailsScriptRunner.java:378)
	at org.codehaus.groovy.grails.cli.GrailsScriptRunner.main(GrailsScriptRunner.java:226)
	at org.codehaus.groovy.grails.cli.support.GrailsStarter.rootLoader(GrailsStarter.java:236)
	at org.codehaus.groovy.grails.cli.support.GrailsStarter.main(GrailsStarter.java:264)
	at org.codehaus.groovy.grails.cli.support.GrailsStarter.rootLoader(GrailsStarter.java:236)
	at org.codehaus.groovy.grails.cli.support.GrailsStarter.main(GrailsStarter.java:264)
```

UserRole file source:

```
package foo.bar.model.security;

import java.io.Serializable;

import javax.persistence.Entity;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;

@Entity @Table(name="user_role")
public class UserRole implements Serializable {

	private static final long serialVersionUID = -7733009390177318889L;

	@Id	
	@ManyToOne(optional=false)
	@JoinColumn(name="user_id")
	private User user;
	
	@Id 
	@ManyToOne(optional=false)
	@JoinColumn(name="role_id")
	private Role role;

	public UserRole() {}
	
	public UserRole(User user, Role role) {
		this.user = user;
		this.role = role;
	}
	
	public User getUser() {
		return user;
	}

	public void setUser(User user) {
		this.user = user;
	}

	public Role getRole() {
		return role;
	}

	public void setRole(Role role) {
		this.role = role;
	}
	
	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof UserRole)) {
			return false;
		}
		
		UserRole other = (UserRole) obj;
		return (user != null 
				&& role != null 
				&& other.user != null
				&& other.role != null
				&&  user.equals(other.user) 
				&&  role.equals(other.role))
			|| super.equals(obj);
	}

    @Override
    public int hashCode() {
	int result = 37;
	if (user != null && user.getId() != null) {
	    result = result*17 + user.getId().intValue();
	}
	if (role != null && role.getId() != null) {
	    result = result*17 + role.getId().intValue();
	}
	return result;
    }
}
```

Test class:

```
package foo.bar.model.security

import grails.plugin.springsecurity.SpringSecurityService
import grails.plugin.springsecurity.acl.AclSid
import grails.plugin.springsecurity.oauth.MailruOAuthToken
import grails.test.mixin.Mock
import grails.test.mixin.TestFor
import grails.test.mixin.web.ControllerUnitTestMixin

import org.scribe.model.Token

import spock.lang.Specification

@Mock([UserRole])
@TestFor(RegisterController)
class RegisterControllerSpec extends Specification {
}
```

**Comment by graemerocher** [comment]
The unit testing implementation doesn't currently support composite identifiers. You can however user `HibernateSpec` to write a unit test instead

**Comment by badbob** [comment]
@graemerocher, looks like there is no such class (_HibernateSpec_) included into Grails 2.5.*. Is this class available by external dependency?

**Comment by graemerocher** [comment]
It is available if you upgrade to GORM 5/6 in Grails 2.5.x

**Comment by badbob** [comment]
@graemerocher, thanks for your answer. Looks like HibernateSpec is available since 6.0 version [http://gorm.grails.org/6.0.x/hibernate/api/grails/test/hibernate/HibernateSpec.html]. So I try to upgrade up to 6.0.7 version and failed. The reason is my domain classes (entities) declared in separate maven project (artifact) and declared in hibernate.cfg.xml. This aproach worked file with hibernate4:3.15, but after update is get this exception:

```
groovy.lang.MissingMethodException: No signature of method: static foo.bar.model.security.User.findByUsername() is applicable for argument types: (java.lang.String) 
```

Looks like hibernate4:6.0.7 ignores hibernate.cfg.xml configuration file.

**Comment by graemerocher** [comment]
yeah 6.1 milestones are required for JPA annotation support. Anyway looks like for this use case you will have to write an integration test.
